### PR TITLE
Add checks to compare some commonly None values in docker_container

### LINKED ISSUE #35070: docker_container does not recreate the container if dns_servers is removed
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
docker_container

##### ANSIBLE VERSION

```
$ ansible --version
ansible 2.4.1.0
  config file = /home/ubuntu/.ansible.cfg
  configured module search path = [u'/home/ubuntu/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.12 (default, Nov 20 2017, 18:23:56) [GCC 5.4.0 20160609]
```

##### CONFIGURATION
ANSIBLE_SSH_ARGS(/home/ubuntu/.ansible.cfg) = -o StrictHostKeyChecking=no
DEFAULT_REMOTE_USER(/home/ubuntu/.ansible.cfg) = hoylu
DEFAULT_ROLES_PATH(/home/ubuntu/.ansible.cfg) = [u'/home/ubuntu/ansible-roles']
DEFAULT_SCP_IF_SSH(/home/ubuntu/.ansible.cfg) = True
DEFAULT_TRANSPORT(/home/ubuntu/.ansible.cfg) = ssh

##### OS / ENVIRONMENT
NA

##### SUMMARY
when a container is started with the 'dns_servers' parameter set, and then the parameter is . removed and the task is run again, one would expect that the container gets re-created with the Docker Dns parameter set to null.

##### STEPS TO REPRODUCE

```yaml
- name: "test docker containerr recreate with server_dns"
  hosts:  localhost
  tasks:
    - name: start the test container
      docker_container:
        name: tester
        image: redis
    - name: this should restart the test container
      docker_container:
        name: tester
        image: redis
        dns_servers: "8.8.8.8"
    - name: this should also restart the test container
      docker_container:
        name: tester
        image: redis
```

##### EXPECTED RESULTS
Expected a `changed` on the task 'this should also restart the container'

##### ACTUAL RESULTS
Got an `ok` on the task. The container was not restarted despite the configuration being different. Running it in -vvvv mode shows `module_args:dns_servers` set to null, but that is not applied to the container.

```
PLAY [test docker containerr recreate with server_dns] ********************************************************************************************************

TASK [Gathering Facts] ****************************************************************************************************************************************
ok: [localhost]

TASK [start the test container] *******************************************************************************************************************************
changed: [localhost]

TASK [this should restart the test container] *****************************************************************************************************************
changed: [localhost]

TASK [this should also restart the test container] ************************************************************************************************************
ok: [localhost]

PLAY RECAP ****************************************************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0
```

### PR DESCRIPTION
**Pull request by bengerman13**
##### SUMMARY
The current docker_container comparison logic does not bother checking any unset values, but these can be potentially interesting. For instance, if you specify no volumes, and the container currently has volumes bound, that is a valid difference that should be updated. 

This PR adds logic to check some parameters of this type. It does this selectively, and the list it checks is probably incomplete.
Fixes #35070

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
docker_container

##### ANSIBLE VERSION

```
ben@munin:~/test$ ansible --version
ansible 2.6.0 (better-docker-container-compare f237911503) last updated 2018/05/16 13:44:57 (GMT -700)
  config file = /home/ben/test/ansible.cfg
  configured module search path = [u'/home/ben/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/ben/dev/github/ansible/ansible/lib/ansible
  executable location = /home/ben/venvs/ansible-devel/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]

```

##### ADDITIONAL INFORMATION
testing playbook:
```
- name: "test docker container recreate"
  hosts:  localhost
  tasks:
    - name: start the test container
      docker_container:
        name: tester
        image: redis
    - name: this should not restart the container
      docker_container:
        name: tester
        image: redis
    - name: this should restart the test container
      docker_container:
        name: tester
        image: redis
        dns_servers: "8.8.8.8"
    - name: this should not restart the test container
      docker_container:
        name: tester
        image: redis
        dns_servers: "8.8.8.8"
    - name: this should also restart the test container
      docker_container:
        name: tester
        image: redis
    - name: this should not restart the test container
      docker_container:
        name: tester
        image: redis
```

### CODE REVIEW

**Comment by bengerman13** [comment]
bot_status

**Comment by bengerman13** [comment]
ready_for_review

**Comment by kassiansun** [comment]
> For instance, if you specify no volumes, and the container currently has volumes bound, that is a valid difference that should be updated.

This was previously discussed. IMO, `docker_container` should touch remote host minimally. Take volume as a example, the volume previously mounted should not affect the container's function, so we should not restart it, which will interrupt the services.

**Comment by bengerman13** [comment]
Thanks for taking a look at this.

I just grabbed this as a ticket to be worked on that affects a module I use a lot - I don't know that I've run into this issue directly myself, but I can imagine some cases where this change would be nice. That is to say, I worked on it because it seems valuable, but I don't have much of a dog in this fight. @bjornicus may have more to say

The rationale I see for this change is that without it, there is not an easy way to remove all volume mounts or remove all DNS servers, etc. There are definitely cases where a mounted volume affects a container's function. I frequently mount volumes that shadow directories in the image, for instance conf.d-style directories.

Would it make sense to add  `purge_*` flags, much like several of the AWS modules do (`purge_origins` in `cloudfront_distribution`)?

**Comment by kassiansun** [comment]
@bengerman13 Right, add another flag like `strict` or `purge` may be a suitable solution.

File: lib/ansible/modules/cloud/docker/docker_container.py [hunk 1]
```diff
@@ -1269,47 +1269,67 @@ def has_different_configuration(self, image):
...
-                    # no match. record the differences
-                    item = dict()
-                    item[key] = dict(
-                        parameter=getattr(self.parameters, key),
-                        container=value
-                    )
-                    differences.append(item)
+            elif key in interesting_unset_params and current_value:
+                self.log('%s is None in parameters, and %s in running container' % (key, str(current_value)))
+                match = False
```
Comment (kassiansun) [inline lib/ansible/modules/cloud/docker/docker_container.py:1323]: Add a parameter `strict_mode` and set `match = False` conditionally at here?

**Comment by kassiansun** [comment]
After some struggling with `docker_container` recently, adding a mechanism for detecting the "equality (this patch)" instead of "included (current implementation)" is necessary.

File: lib/ansible/modules/cloud/docker/docker_container.py [hunk 2]
```diff
@@ -1269,47 +1269,67 @@ def has_different_configuration(self, image):
...
+
+            if not match:
+                # no match. record the differences
+                item = dict()
+                item[key] = dict(
+                    parameter=getattr(self.parameters, key),
```
Comment (kassiansun) [inline lib/ansible/modules/cloud/docker/docker_container.py:1329]: replace this with `desired_value` like other codes?

**Comment by felixfontein** [comment]
At least for volumes, this can be solved by what we are discussing in #43679 and by specifying an empty volume list as a `docker_container` option -- as opposed to not specifying one at all. Not specifying something means "don't change the container with regard to this if it is already running", but specifying an empty list can be interpreted as "remove all volumes" or "add no new volumes" depending on the new setting proposed in #43679 (at least in the discussion for now).

**Comment by felixfontein** [comment]
After working with `docker_container` for some time now, I don't think this is the way to go. The idea is that not specified options should not have an impact on deciding whether the container should be restarted/recreated/updated. If you want to ensure that something is *not* there, you should specify that explicitly by giving an empty list of volumes, shorter list of DNS servers, etc. Unfortunately, also that doesn't work for now since `docker_container` doesn't have a "strict mode". There's #44789 now, though, which implements a fine-grained "strict mode" where you can say per option whether (simplified) you want strict comparison or not.

**Comment by kassiansun** [comment]
duplicate_of #44789

**Pull request closed without merging**